### PR TITLE
docs(hooks): useCopyToClipboard

### PR DESCRIPTION
**Pull request by mainframev**
Slight improvment. `react-copy-to-clipboard` lib replaced with `copy-to-clipboard` and hook. Now it works even better, state changes back after `3000ms`, for instance, in the Palette component state changes back to `HEX` after `COPIED`.

 Orbit.kiwi: https://orbit-docs-feat-useCopytoClipboard.surge.sh
 Storybook: https://orbit-feat-useCopytoClipboard.surge.sh

### CODE REVIEW

File: docs/src/hooks/useCopyToClipboard.ts [hunk 1]
```diff
@@ -0,0 +1,29 @@
+// @flow
+import { useEffect, useState, useCallback } from "react";
+import copy from "copy-to-clipboard";
+
+const useCopyToClipboard = (
+  timer: number | undefined = 3000,
```
Comment (silvenon) [inline docs/src/hooks/useCopyToClipboard.ts:6]: Why is 2s being increased to 3s? The PR description doesn't say.

File: docs/src/components/VisualList.tsx [hunk 2]
```diff
@@ -69,10 +65,7 @@ const Visual = ({
   smallVisual,
   visualName,
 }: VisualProps) => {
-  const [copied, setCopied] = useState(false);
-  useEffect(() => {
-    copyTimeout(copied, setCopied);
-  }, [copied, setCopied]);
+  const [copied, setCopied] = useCopyToClipboard();
```
Comment (silvenon) [inline docs/src/components/VisualList.tsx:68]: I would rename `setCopied` to `copy` or `copyToClipboard` instead, to suggest a text argument rather than boolean. (And rename it elsewhere, too, not only here.)

**Review by silvenon — CHANGES_REQUESTED** [review]
You can use [timer mocks](https://jestjs.io/docs/timer-mocks) to add tests.

File: package.json [hunk 3]
```diff
@@ -132,14 +132,14 @@
...
-    "react": "^16.14.0",
     "react-dom": "^16.14.0",
     "react-element-to-jsx-string": "^14.3.2",
     "react-test-renderer": "^16.14.0",
+    "react": "^16.14.0",
     "remark-cli": "^9.0.0",
     "remark-frontmatter": "^2.0.0",
-    "remark-lint": "^8.0.0",
     "remark-lint-first-heading-level": "^2.0.1",
+    "remark-lint": "^8.0.0",
```
[the remark this answers is not in the document]
Comment (mainframev) [reply package.json:142]: arranged in asc order

File: docs/src/hooks/useCopyToClipboard.ts (hunk 1 above)
Comment (mainframev) [reply docs/src/hooks/useCopyToClipboard.ts:6]: 2s seemed as a too fast for me.

File: package.json (hunk 3 above)
[the remark this answers is not in the document]
Comment (silvenon) [reply package.json:142]: Manually? Won't they get automatically moved back by package manager?

[the remark this answers is not in the document]
Comment (mainframev) [reply package.json:142]: not manually, by extension

[the remark this answers is not in the document]
Comment (mainframev) [reply package.json:142]: they won't

[the remark this answers is not in the document]
Comment (silvenon) [reply package.json:142]: I tested it by installing a dependency, and this change got reversed. Package managers enforce a certain order, there's a reason dependencies were all ordered a certain way, so I recommend disabling your extension.

[the remark this answers is not in the document]
Comment (silvenon) [reply package.json:142]: Or aligning it to act the same way as package managers.

File: docs/src/hooks/useCopyToClipboard.ts [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+      copy(text);
+      setCopied(true);
+    } else {
+      setCopied(false);
+      console.error(`Cannot copy typeof ${text} to clipboard`);
+    }
+  }, []);
+
+  useEffect(() => {
+    let timeout;
```
[the remark this answers is not in the document]
Comment (mainframev) [reply docs/src/hooks/useCopyToClipboard.ts:21]: having it's inside useEffect is ok, i think

File: package.json (hunk 3 above)
[the remark this answers is not in the document]
Comment (mainframev) [reply package.json:142]: that's not automatically, no need for disable it. Keep calm)

File: docs/src/hooks/useCopyToClipboard.ts (hunk 1 above)
Comment (silvenon) [reply docs/src/hooks/useCopyToClipboard.ts:6]: Fine by me, just to the description that 3s was a personal choice because it feels better.

Comment (mainframev) [reply docs/src/hooks/useCopyToClipboard.ts:6]: btw, PR description had `3000ms` :)

**Comment by mainframev** [comment]
> You can use [timer mocks](https://jestjs.io/docs/timer-mocks) to add tests.

Thanks, I know. But I did not plan to write tests for it. You think I should add that?

**Comment by CollierCZ** [comment]
How is this an improvement? I'm not against it, but I'm not sure I understand the benefit. The button text in Palette already switched back to the original. I liked it at 2s, but that's just a preference. Is there a performance or other benefit?

**Comment by silvenon** [comment]
> I did not plan to write tests for it. You think I should add that?

Yep, because it has non-trivial behavior so it's good to ensure edge cases, but considering that this is not critical logic (i.e. not in `@kiwicom/orbit-components`), it remains only a suggestion, you don't have to. 😉

> I'm not sure I understand the benefit

It's better in multiple ways. This functionality has two pieces, managing state which is used to notify the user that the text has been copied to clipboard, and the actual copying to clipboard. These two related pieces were separated before this change, and this hook unified them. Also, `react-copy-to-clipboard` didn't have any other output, it was only accessing its child to implement copying to clipboard on click, which is not only limiting, it also hides what's going on. React components should not be doing too much to its children other than simply outputting them or iterating through them for whatever reason. This behavior was better implemented as a hook.

**Comment by CollierCZ** [comment]
> it also hides what's going on

This I don't see. At least, not in a way that seems more open now. The rest seems like a minor benefit that would be nice to think about after we do the things that are blocking a beta release.

**Comment by mainframev** [comment]
> > it also hides what's going on
> 
> This I don't see. At least, not in a way that seems more open now. The rest seems like a minor benefit that would be nice to think about after we do the things that are blocking a beta release. But maybe just me. 🤷

I think it's good to solve it now. It's always happen that `we can solve it later` and then it will be forgotten, plus i am working on that small improvement and you can keep doing more urgent stuff :)

**Comment by silvenon** [comment]
> The rest seems like a minor benefit that would be nice to think about after we do the things that are blocking a beta release.

I totally get that, the benefits indeed were secondary compared to other priorities, but occasionally I think it's good to warm up for work with non-critical tasks, I doubt that it took him a long time.

Viktor and I just talked about his next task for beta release, so we're on track.

File: docs/src/hooks/__tests__/useCopyToClipboard.test.ts [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+    const { result } = renderHook(() => useCopyToClipboard());
+
+    expect(result.current[0]).toBe(false);
+
+    act(() => {
+      jest.useFakeTimers();
+      result.current[1]("kek");
+    });
+
+    expect(result.current[0]).toBe(true);
```
Comment (silvenon) [inline docs/src/hooks/__tests__/useCopyToClipboard.test.ts:16]: The timeout is missing from the test, otherwise all good. 👍 

```ts
    act(() => {
      jest.runAllTimers();
    });

    expect(result.current[0]).toBe(false);
```

Also, I'm not sure how CI passes without mocking `copy-to-clipboard` (because of that `window.prompt` error), so you also need this:

```ts
jest.mock("copy-to-clipboard", () => () => {});
```
